Fixes openmpi configure args for multinode execution on slurm

tbhaxor opened the pull request:
- disable btl-uct with ucx for v4+
- actually internal configuration of pmix and hwloc based on variant

tbhaxor committed:
fix (packages): openmpi configure args

- disable btl-uct with ucx for v4+
- actually internal configuration of pmix and hwloc based on variant

hpcnpatel commented:
Hello, and thank you for this contribution!

I'd like to emphasize that this PR is helpful as it assists in addressing an issue we've been facing with OpenMPI. Looking forward to the merge

alalazo commented:
@spackbot run pipeline

alalazo commented:
@hppritcha Do you have some time to review this change?

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -995,6 +1005,12 @@ def configure_args(self):
             self.enable_or_disable("mpi-thread-multiple", variant="thread_multiple")
         )
 
+        # With OpenMPI 4.0 and above, there could be compilation errors from “btl_uct” component.
+        # This component is not critical for using UCX; so it could be disabled.
+        # See https://openucx.readthedocs.io/en/master/running.html#id1
+        if "^ucx" in spec and spec.satisfies("@4:"):
+            config_args.append("--enable-mca-no-build=btl-uct")
+
```
Did you actually run into an issue?

I didn't read the full thread, but looks like it was resolved: https://github.com/open-mpi/ompi/issues/6666

tbhaxor replied:
Actually I had the issue on multinode slurm. So the Admins of our HPC platform included this and it worked. 

Also I would like to mention it was in v4.0.5 of openmpi. So I have scheduled a job. I will update the PR after it has ran.

hppritcha commented:
@tbhaxor are you done with pushing commits, i.e. is this PR ready for review?

tbhaxor commented:
> @tbhaxor are you done with pushing commits, i.e. is this PR ready for review?

Please wait for sometime. I have submitted the job on our SLURM. I will inform you here when everything looks good to me.

tbhaxor committed:
fix (packages): simplify configure_args of openmpi

Remove `--enable-mca-no-build=btl-uct` check for ucx spec
as it is already fixed in the openmpi open-mpi/ompi#6666

tbhaxor commented:
> @tbhaxor are you done with pushing commits, i.e. is this PR ready for review?

Yes, now it is ready for review.

tbhaxor committed:
Merge branch 'develop' into fix/openmpi
